File: Dart/src/Direction.dart
class Direction {
  final int dx;
  final int dy;
  Direction(this.dx, this.dy);
  
  Direction rotate(int rotation) {
    switch(rotation) {
      case Rotation.LEFT: 
        return new Direction(-dy, dx);
      case Rotation.RIGHT: 
        return new Direction(dy, -dx);
      default:
        throw 'Error';
    }
  }
  
  Cell moveFrom(Cell fromCell) {
    return new Cell(fromCell.x + dx, fromCell.y + dy);
  }
  
  bool operator == (Direction other) => other!=null && other.dx==dx && other.dy==dy;
  
  int hashCode() => dx + dy;
  
  factory Direction.north() {
    return new Direction(0, 1);
  }
  
  factory Direction.east() {
    return new Direction(1, 0);
  }
  
  factory Direction.south() {
    return new Direction(0, -1);
  }
  
  factory Direction.west() {
    return new Direction(-1, 0);
  }
}

class Rotation {
  static final int LEFT = 0;
  static final int RIGHT = 1;
}

File: Kotlin/src/Direction.kt
class Direction(val dx: Int, val dy: Int) {

    class object {
        val North = Direction(0, 1);
        val East = Direction(1, 0);
        val South = Direction(0, -1);
        val West = Direction(-1, 0);
    }

    fun rotate(rotation: Int) : Direction {
        when(rotation) {
            Turn.Left -> return Direction(-dy, dx)
            Turn.Right -> return Direction(dy, -dx)
            else -> throw Exception("Not implemented")
        }
    }

    fun moveFrom(from: Position) : Position {
        return Position(from.x + dx, from.y + dy);
    }

    fun equals(obj : Any?) : Boolean {
        if(obj is Direction) {
            return dx==obj.dx && dy==obj.dy
        } else {
            return false
        }
    }

    fun hashCode() : Int {
        return dx + dy
    }
}

class Turn() {
    class object {
        val Left = 0
        val Right = 1
    }
}